### PR TITLE
Add schema watch support in Spanner datastore driver

### PR DESCRIPTION
**Pull request by josephschorr**

### CODE REVIEW

File: internal/datastore/spanner/watch.go [hunk 1]
```diff
@@ -19,8 +23,21 @@ import (
 
 const (
 	RelationTupleChangeStreamName = "relation_tuple_stream"
+	SchemaChangeStreamName        = "schema_change_stream"
 )
 
+var retryHistogram = prometheus.NewHistogram(prometheus.HistogramOpts{
```
Comment (jzelinskie) [inline internal/datastore/spanner/watch.go:29]: Why not have a single metric if both crdb and cockroach are using this?

Comment (josephschorr) [reply internal/datastore/spanner/watch.go:29]: We could, but then it has to be defined in a central place, and that doesn't currently impact the other datastores

File: internal/datastore/spanner/watch.go [hunk 2]
```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
...
+	updates := make(chan *datastore.SchemaState, 10)
+	errs := make(chan error, 1)
+
+	lastRevision := afterRevision
+	retryCount := -1
+	go (func() {
+		running := true
+		for running {
+			retryCount++
+			retryHistogram.Observe(float64(retryCount))
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:197]: Why not a gauge or a count(1)? This number is basically monotonically increasing over time, what kind of graph were you planning to plot?

File: internal/datastore/spanner/watch.go [hunk 3]
```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
...
+	afterRevision := afterRevisionRaw.(revision.Decimal)
+	reader, err := changestreams.NewReaderWithConfig(
+		ctx,
+		project,
+		instance,
+		database,
+		SchemaChangeStreamName,
+		changestreams.Config{
+			StartTimestamp:    timestampFromRevision(afterRevision),
+			HeartbeatInterval: 100 * time.Millisecond, // 100ms is the minimum allowed
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:244]: N spicedb replicas running 10 heartbeats per second could be a bunch of heartbeats. Do we want to be more conservative, or perhaps expose this in a parameter now?

File: internal/datastore/spanner/watch.go [hunk 4]
```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
...
+
+func (sd spannerDatastore) watchSchemaWithoutRetry(ctx context.Context, afterRevisionRaw datastore.Revision, processUpdate func(update *datastore.SchemaState, err error)) {
+	project, instance, database, err := parseDatabaseName(sd.database)
+	if err != nil {
+		processUpdate(nil, err)
+		return
+	}
+
+	afterRevision := afterRevisionRaw.(revision.Decimal)
+	reader, err := changestreams.NewReaderWithConfig(
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:236]: is there a priority associated with change streams? should we put it by default to low priority so it does not disrupt the critical path?

File: internal/datastore/spanner/watch.go [hunk 5]
```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
 	}
 	return nil, nil
 }
+
+const maxSchemaWatchRetryCount = 15
+
+func (sd spannerDatastore) WatchSchema(ctx context.Context, afterRevision datastore.Revision) (<-chan *datastore.SchemaState, <-chan error) {
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:187]: can we add mocked tests at least if the emulator does not properly do changestreams?

File: internal/datastore/spanner/watch.go [hunk 6]
```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
...
+
+	err = reader.Read(ctx, func(result *changestreams.ReadResult) error {
+		// See: https://cloud.google.com/spanner/docs/change-streams/details
+		for _, record := range result.ChangeRecords {
+			for _, dcr := range record.DataChangeRecords {
+				changeRevision := revisionFromTimestamp(dcr.CommitTimestamp)
+				modType := dcr.ModType // options are INSERT, UPDATE, DELETE
+
+				for _, mod := range dcr.Mods {
+					primaryKeyColumnValues := mod.Keys.Value.(map[string]any)
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:263]: could this type assertion panic? maybe do it the safe way with `ok`?

**Review by vroldanbet — COMMENTED** [review]
This looks good to me, I left a bunch of questions. I think we need to add some basic unit tests, even if it means mocking

File: internal/datastore/spanner/watch.go [hunk 7]
```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
...
+
+const maxSchemaWatchRetryCount = 15
+
+func (sd spannerDatastore) WatchSchema(ctx context.Context, afterRevision datastore.Revision) (<-chan *datastore.SchemaState, <-chan error) {
+	updates := make(chan *datastore.SchemaState, 10)
+	errs := make(chan error, 1)
+
+	lastRevision := afterRevision
+	retryCount := -1
+	go (func() {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/datastore/spanner/watch.go:193]: I personally find it more readable but removed

File: internal/datastore/spanner/watch.go (hunk 4 above)
Comment (josephschorr) [reply internal/datastore/spanner/watch.go:236]: Added a low priority

File: internal/datastore/spanner/watch.go (hunk 5 above)
Comment (josephschorr) [reply internal/datastore/spanner/watch.go:187]: As discussed, will wait for the emulator's support to catch up

**Comment by josephschorr** [comment]
Updated

File: pkg/cmd/datastore/datastore.go [hunk 8]
```diff
@@ -210,6 +213,8 @@ func RegisterDatastoreFlagsWithPrefix(flagSet *pflag.FlagSet, prefix string, opt
 	flagSet.StringVar(&opts.MigrationPhase, flagName("datastore-migration-phase"), "", "datastore-specific flag that should be used to signal to a datastore which phase of a multi-step migration it is in")
 	flagSet.Uint16Var(&opts.WatchBufferLength, flagName("datastore-watch-buffer-length"), 1024, "how many events the watch buffer should queue before forcefully disconnecting reader")
 
+	flagSet.DurationVar(&opts.SchemaWatchHeartbeat, flagName("datastore-schema-watch-heartbeat"), 0*time.Millisecond, "heartbeat time on the schema watch in the datastore (if supported). 0 means to default to the datastore's minium.")
```
Comment (vroldanbet) [inline pkg/cmd/datastore/datastore.go:216]:
```suggestion
	flagSet.DurationVar(&opts.SchemaWatchHeartbeat, flagName("datastore-schema-watch-heartbeat"), 0*time.Millisecond, "heartbeat time on the schema watch in the datastore (if supported). 0 means to default to the datastore's minimum.")
```

**Review by vroldanbet — APPROVED** [review]

File: pkg/cmd/datastore/datastore.go (hunk 8 above)
Comment (josephschorr) [reply pkg/cmd/datastore/datastore.go]: Fixed

File: internal/datastore/spanner/watch.go (hunk 1 above)
Comment (jzelinskie) [reply internal/datastore/spanner/watch.go:31]: Defining it in a central places seems pretty minor compared to users having to dashboard two different metrics.

**Pull request merged**